### PR TITLE
Fix/targets

### LINKED ISSUE #104: You are not able to defend an enemy
When you play a Defend card and target an enemy the block power is added to the player.

Expected => the block power is added to the target, player or enemy.

Why ? 

1 - you can strike the enemy and the player so you should be able to do the same for defending
2 - in the future the game could include cards like "gain as much health as the enemy has blocks" thus you would want to stack the defend card on the enemy

### LINKED ISSUE #106: Using Bash card on player throw an error
When you target the player with a Bash card it do nothing and the card stays hang. The console throw an error message : 

```
index.js:42 Error: TypeError: Cannot read property 'currentHealth' of undefined
    at Object.dequeue (action-manager.js:26)
    at Object.dequeue (index.js:39)
    at App.dequeue (app.js:50)
    at onComplete (app.js:63)
    at App.playCard (app.js:65)
    at onRelease (dragdrop.js:41)
    at Lt (Draggable.js:294)
    at HTMLDocument.t (Draggable.js:2126)
```

Screen Shot 2020-08-19 at 16 28 35

I think this issue can be delayed until a solution has been found for [issue 104 / You are not able to defend an enemy](https://github.com/oskarrough/slaytheweb/issues/104#issue-681803800)

### PR DESCRIPTION
**Pull request by oskarrough**
All cards have a `target` property that specifies which targets the card can be played at. Allowed ones are currently `player`, `enemy` and `all enemies`.

Before this PR, it wasn't really enforced e.g. a "Strike" card with `target: enemy` could also be played on yourself, the player.

This PR changes that and makes sure the `card.target` is respected.

If we later want to allow attacking yourself, we can now do that in a cleaner way by creating a new target `player and enemies` or something like that. But that's for later, at least now it's "clean".

Closes #104 , fixes #106.

### CODE REVIEW

**Comment by oskarrough** [comment]
@SamuelRott thoughts? This overlaps a bit with your new `conditions` logic maybe?

File: public/game/cards.js [hunk 1]
```diff
@@ -54,7 +54,7 @@ export const cards = [
 		energy: 1,
 		damage: 5,
 		block: 5,
-		target: 'self and enemy',
```
Comment (oskarrough) [inline public/game/cards.js:57]: this wasn't even implemented

File: public/game/utils.js [hunk 2]
```diff
@@ -59,6 +59,10 @@ export function getTargets(state, target) {
 	throw new Error(`Can not find monster with target: "${target}"`)
 }
 
+export function canPlayCard(card, target) {
```
Comment (oskarrough) [inline public/game/utils.js:62]: rename to `cardHasValidTarget`?

File: public/ui/dragdrop.js [hunk 3]
```diff
@@ -34,14 +36,24 @@ export default function enableDragDrop(container, onAdd) {
...
-				if (targetElement) {
-					const cardElement = this.target
-					const index = Array.from(targetElement.parentNode.children).indexOf(targetElement)
-					const target = targetElement.dataset.type + index
-					onAdd(cardElement.dataset.id, target, cardElement)
-				} else {
-					gsap.to(this.target, {x: this.startX, y: this.startY})
+
+				if (!targetElement) {
+					return gsap.to(this.target, {x: this.startX, y: this.startY})
```
Comment (oskarrough) [inline public/ui/dragdrop.js:41]: move this animation into a function since we reuse it below

File: public/ui/dragdrop.js [hunk 4]
```diff
@@ -34,14 +36,24 @@ export default function enableDragDrop(container, onAdd) {
...
+				const target = targetElement.dataset.type + targetIndex
+
+				// Check if the card allows us to drop it here.
+				const cardName = cardElement.querySelector('.Card-title').textContent
+				const card = cards.find(c => c.name === cardName)
+				if (!canPlayCard(card, target)) {
+					return gsap.to(this.target, {x: this.startX, y: this.startY})
 				}
+
+				onAdd(cardElement.dataset.id, target, cardElement)
```
Comment (oskarrough) [inline public/ui/dragdrop.js:55]: rename `onAdd` to `afterRelease`?

File: public/game/utils.js (hunk 2 above)
Comment (SamuelRott) [reply public/game/utils.js:62]: Yes, It would make it easier to get it.

**Comment by SamuelRott** [comment]
Hello, 

I dont think this overlap the conditions logic. They both concern different part of the game and rely on different logic.

For the long run I would say that card.target is not going to be enough. We are going to need card.allowedTarget = [].
Where card.target is defined only if the card has a specific target and card.allowedTarget is a list of the possible target of this card (we currently don't need this).

**Pull request merged**